src/front/domain_socket.rs

```diff
@@ -32,25 +38,35 @@ pub struct DomainSocketListener {
 
 impl DomainSocketListener {
     /// Initialise the connection to the Unix socket.
-    ///
-    /// # Panics
-    /// - if a file/socket exists at the path specified for the socket and `remove_file`
-    /// fails
-    /// - if binding to the socket path fails
     pub fn new(timeout: Duration) -> Result<Self> {
-        // If this Parsec instance was socket activated (see the `parsec.socket`
+        #[cfg(feature = "parsec-user-and-clients-group")]
+        DomainSocketListener::check_user_details()?;
+
+        // is Parsec instance was socket activated (see the `parsec.socket`
         // file), the listener will be opened by systemd and passed to the
         // process.
         // If Parsec was service activated or not started under systemd, this
         // will return `0`.
         let listener = match sd_notify::listen_fds()? {
             0 => {
                 let socket = Path::new(SOCKET_PATH);
-
-                if socket.exists() {
-                    fs::remove_file(&socket)?;
+                if let Some(parent_dir) = socket.parent() {
+                    if !parent_dir.exists() {
+                        fs::create_dir_all(parent_dir)?;
+                        #[cfg(feature = "parsec-user-and-clients-group")]
+                        DomainSocketListener::set_socket_dir_permissions(parent_dir)?;
+                    } else if socket.exists() {
+                        fs::remove_file(&socket)?;
+                    }
+                } else {
+                    // Invalid path - socket should not be created in root
+                    // `socket.path` returns None if there is no parent directory (i.e. is in root)
+                    error!("Socket path {} is invalid.", SOCKET_PATH);
+                    return Err(Error::new(
+                        ErrorKind::InvalidInput,
+                        "Socket path is invalid",
+                    ));
                 }
```

Comment:
Is this part actually needed, since the path is declared statically at the top?

Reply:
I don't think so. It was added because `socket.parent()` _can_ fail, but should never in our case. Does it make sense to remove the `else`? It should never be hit, but then if you look at the code without the else, it looks like it would continue as normal, even if it failed. Maybe converting the `Option` that `parent()` returns to a `Result` and returning the error (if there is one - which there never should be!) instead, to make the code more compact?

Reply:
You can simply `unwrap` the result of `socket.parent()` - even if we ever change it to something where that fails, the service won't start so the CI will let us know.